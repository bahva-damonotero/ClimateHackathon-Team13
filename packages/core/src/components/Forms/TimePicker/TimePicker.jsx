import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import timePicker from '@uswds/uswds/js/usa-time-picker';
import comboBox from '@uswds/uswds/js/usa-combo-box';

import { createClassString } from '@boozallen-uip/uswds-react-utils';

import { FormError, Label } from '..';
import { Hint } from '../..';

export default function TimePicker({
  children,
  className,
  disabled,
  error,
  hint,
  id,
  maxTime,
  minTime,
  onChange,
  placeholder,
  step,
  use24Hour,
  value
}) {
  const containerRef = useRef();
  const timePickerRef = useRef();

  // initialization
  useEffect(() => {
    const containerEl = containerRef.current;
    if (!containerEl) return;

    // call USWDS's `on` functions to initialize the component
    timePicker.on(containerEl);
    comboBox.on(containerEl);

    // hook into the select's `change` event to monitor changes
    const selectEl = containerEl.querySelector('select');
    selectEl.addEventListener('change', (e) => {
      if (onChange) onChange(e);
    });

    // 24hr format
    if (use24Hour) {
      const options = containerEl.querySelectorAll('option');
      for (const option of options) {
        option.innerHTML = option.value;
      }
    }

    // on dismount, call USWDS's `off` functions to cleanup the component
    return () => {
      comboBox.off(containerEl);
      timePicker.off(containerEl);
    };
  }, []);

  // update the time picker's value whenever the value changes outside
  // of this component
  useEffect(() => {
    const timePickerEl = timePickerRef.current;
    if (!timePickerEl) return;

    const selectEl = timePickerEl.querySelector('select');
    const inputEl = timePickerEl.querySelector(`#${id}`);

    if (value !== selectEl.value) {
      selectEl.value = value;

      const valueOption = selectEl.querySelector(`option[value="${value}"]`);
      if (valueOption) {
        inputEl.value = valueOption.innerHTML;
      }
    }
  }, [value]);

  // call USWDS's `enable` or `disable` function when the disabled prop changes
  useEffect(() => {
    const timePickerEl = timePickerRef.current;
    if (!timePickerEl) return;

    const disabledState = timePickerEl.querySelector(`#${id}`).disabled;
    if (disabled === disabledState) return; // the input already has the right disabled state

    if (disabled) comboBox.disable(timePickerEl);
    else comboBox.enable(timePickerEl);
  }, [disabled]);

  const labelId = `${id}--label`;
  const hintId = `${id}--hint`;

  return (
    <div
      className={createClassString(['bah-uswds-time-picker', className])}
      ref={containerRef}
    >
      <Label id={labelId} target={id}>
        {children}
      </Label>

      {hint && <Hint id={hintId}>{hint}</Hint>}

      <FormError error={error} />

      <div
        className="usa-time-picker"
        data-default-value={value}
        data-max-time={maxTime}
        data-min-time={minTime}
        data-placeholder={placeholder}
        data-step={step}
        ref={timePickerRef}
      >
        <input
          aria-describedby={createClassString([labelId, hint ? hintId : null])}
          className="usa-input"
          disabled={disabled}
          id={id}
          name={id}
          type="text"
        />
      </div>
    </div>
  );
}

TimePicker.propTypes = {
  /** Label for input */
  children: PropTypes.node.isRequired,
  /** Any additional classes to apply */
  className: PropTypes.string,
  /** Should the time picker be disabled? */
  disabled: PropTypes.bool,
  /** Error message to display */
  error: PropTypes.string,
  /** Hint to be shown for the time picker */
  hint: PropTypes.node,
  /** ID of the input */
  id: PropTypes.string.isRequired,
  /** The end time used in the time picker in `hh:mm` 24-hour format. The default is `23:59` */
  maxTime: PropTypes.string,
  /** The start time used in the time picker in `hh:mm` 24-hour format. The default is `00:00` */
  minTime: PropTypes.string,
  /** Function called whenever the time picker value is changed */
  onChange: PropTypes.func,
  /** Text to show when a value has not been selected */
  placeholder: PropTypes.string,
  /**
   * The number of minutes between options.
   * The minimum is 1 minute and the default is 30 minutes.
   */
  step: PropTypes.number,
  /** If `true`, the 24-hour time format is used */
  use24Hour: PropTypes.bool,
  /** Used to set the value of the input. Must be in `hh:mm` 24-hour format */
  value: PropTypes.string
};

TimePicker.defaultProps = {
  disabled: false,
  use24Hour: false
};
