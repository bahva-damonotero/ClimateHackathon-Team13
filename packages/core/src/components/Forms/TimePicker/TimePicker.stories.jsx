import React from 'react';
import { action } from '@storybook/addon-actions';

import TimePicker from '.';
import { Form } from '..';

export default {
  title: 'Forms/TimePicker',
  component: TimePicker
};

const Template = (args) => (
  <Form>
    <TimePicker {...args} />
  </Form>
);

const defaultArgs = {
  children: 'Label',
  id: 'timepicker-example',
  onChange: action('onChange')
};

export const def = Template.bind({});
def.storyName = 'Default';
def.args = {
  ...defaultArgs
};

export const hint = Template.bind({});
hint.args = {
  ...defaultArgs,
  hint: 'hh:mm'
};

export const placeholder = Template.bind({});
placeholder.args = {
  ...defaultArgs,
  placeholder: 'Select a time'
};

export const initialValue = Template.bind({});
initialValue.args = {
  ...defaultArgs,
  value: '13:30'
};

export const minMax = Template.bind({});
minMax.storyName = 'Minimum & Maximum time';
minMax.args = {
  ...defaultArgs,
  maxTime: '17:00',
  minTime: '09:00'
};

export const differentStep = Template.bind({});
differentStep.args = {
  ...defaultArgs,
  step: 15
};

export const twentyFourHour = Template.bind({});
twentyFourHour.storyName = '24-Hour Format';
twentyFourHour.args = {
  ...defaultArgs,
  use24Hour: true
};
