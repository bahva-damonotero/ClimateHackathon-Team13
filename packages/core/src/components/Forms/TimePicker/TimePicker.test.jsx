import React from 'react';
import { createEvent, fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import comboBox from '@uswds/uswds/js/usa-combo-box';
import timePicker from '@uswds/uswds/js/usa-time-picker';

import TimePicker from '.';

describe('TimePicker', () => {
  const id = 'timepicker-test';
  const label = 'Label';

  const INPUT_SELECTOR = `#${id}.usa-combo-box__input`;
  const SELECT_SELECTOR = `.usa-combo-box__select`;

  // Whether using the 24-hour or 12-hour time format, the value
  // for the options is always in 24-hour format. This test asserts
  // that the options' display text matches their values.
  it('should use the 24-hour time format', () => {
    const { container } = render(
      <TimePicker id={id} use24Hour>
        {label}
      </TimePicker>
    );

    const options = container.querySelectorAll('option');
    for (const option of options) {
      expect(option.value === option.innerHTML);
    }
  });

  it('should call `comboBox.on()` and `timePicker.on()` to initialize the component', () => {
    const comboBoxOnSpy = jest.spyOn(comboBox, 'on');
    const timePickerOnSpy = jest.spyOn(timePicker, 'on');

    render(<TimePicker id={id}>{label}</TimePicker>);

    expect(comboBoxOnSpy).toHaveBeenCalled();
    expect(timePickerOnSpy).toHaveBeenCalled();
  });

  it('should call `onChange` when the value changes', () => {
    const onChange = jest.fn();

    const { container } = render(
      <TimePicker id={id} onChange={onChange}>
        {label}
      </TimePicker>
    );

    const selectNode = container.querySelector(SELECT_SELECTOR);

    const event = createEvent(
      'change',
      selectNode,
      {
        detail: {
          value: 0
        }
      },
      { EventType: 'CustomEvent' }
    );
    fireEvent(selectNode, event);

    expect(onChange).toHaveBeenCalledWith(event);
  });

  it('should be able to update the value programmatically', () => {
    // render the time picker with no value
    const initialProps = {
      id,
      onChange: () => {}
    };

    const { container, rerender } = render(
      <TimePicker {...initialProps}>{label}</TimePicker>
    );

    const inputNode = container.querySelector(INPUT_SELECTOR);
    expect(inputNode.value).toBe('');

    // update the time picker's value programmatically/externally
    rerender(
      <TimePicker {...initialProps} value="22:30">
        {label}
      </TimePicker>
    );

    expect(inputNode.value).toBe('10:30pm');
  });

  it('changing the `disabled` prop to true should call `comboBox.disable()`', () => {
    const disableSpy = jest.spyOn(comboBox, 'disable');

    const { rerender } = render(<TimePicker id={id}>{label}</TimePicker>);

    rerender(
      <TimePicker id={id} disabled>
        {label}
      </TimePicker>
    );

    expect(disableSpy).toHaveBeenCalled();
  });

  it('changing the `disabled` prop to false should call `comboBox.enable()`', () => {
    const enableSpy = jest.spyOn(comboBox, 'enable');

    const { rerender } = render(
      <TimePicker id={id} disabled>
        {label}
      </TimePicker>
    );

    rerender(<TimePicker id={id}>{label}</TimePicker>);

    expect(enableSpy).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    const tree = renderer
      .create(
        <TimePicker hint="hh:mm" id={id}>
          {label}
        </TimePicker>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
