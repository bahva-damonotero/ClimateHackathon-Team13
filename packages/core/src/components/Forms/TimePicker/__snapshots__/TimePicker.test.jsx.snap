// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker should match snapshot 1`] = `
<div
  className="bah-uswds-time-picker"
>
  <label
    className="usa-label"
    htmlFor="timepicker-test"
    id="timepicker-test--label"
  >
    Label
  </label>
  <span
    className="usa-hint"
    id="timepicker-test--hint"
  >
    hh:mm
  </span>
  <div
    className="usa-time-picker"
  >
    <input
      aria-describedby="timepicker-test--label timepicker-test--hint"
      className="usa-input"
      disabled={false}
      id="timepicker-test"
      name="timepicker-test"
      type="text"
    />
  </div>
</div>
`;
